### DISCUSSION TITLE
One to one raises not null instead of unique constraint error when creating two children with same parent

### DISCUSSION THREAD

**Discussion by lafrech** [sqlalchemy/sqlalchemy#10448] [Usage Questions]
Trying to implement a 1x1 relation as per [the docs](https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#one-to-one):

```py
class Parent(Base):
    __tablename__ = "parent_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    child: Mapped["Child"] = relationship(back_populates="parent")


class Child(Base):
    __tablename__ = "child_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    parent_id: Mapped[int] = mapped_column(ForeignKey("parent_table.id"))
    parent: Mapped["Parent"] = relationship(back_populates="child")
```

The docs read:

> the Child.parent side of the above relationship remains as a “many-to-one” relationship and is unchanged, and there is no intrinsic system within the ORM itself that prevents more than one Child object to be created against the same Parent during persistence. Instead, techniques such as [unique constraints](https://docs.sqlalchemy.org/en/20/core/constraints.html#schema-unique-constraint) may be used in the actual database schema to enforce this arrangement, where a unique constraint on the Child.parent_id column would ensure that only one Child row may refer to a particular Parent row at a time.

Indeed, if I create a parent and two children, I get no error.

```py
parent_1 = model.Parent()
db.session.add(parent_1)

child_1 = model.Child(parent=parent_1)
db.session.add(child_1)

child_2 = model.Child(parent=parent_1)
db.session.add(child_2)
db.session.flush()
```

I do get an error for this if I set a unique constraint on `Child.parent_id`:

```py
class Child(Base):
    __tablename__ = "child_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    parent_id: Mapped[int] = mapped_column(sqla.ForeignKey("parent_table.id"), unique=True)  # <- here
    parent: Mapped["Parent"] = relationship(back_populates="child")
```

Until then, it conforms to what is written in the docs.

Now, if I slightly change the use case to flush in between the creation of the children, I get more surprising results.

```py
parent_1 = model.Parent()
db.session.add(parent_1)

child_1 = model.Child(parent=parent_1)
db.session.add(child_1)
db.session.flush()  # <- here

child_2 = model.Child(parent=parent_1)
db.session.add(child_2)
db.session.flush()
```

Whether or not the unique constraint is added, I get a NOT NULL constraint error.

What happens is that when creating child 2, SQLAlchemy removes parent from child 1, which is why I get the NOT NULL constraint issue on child 1 (because parent is not optional/nullable).

---------------------------------------

I'm not sure what I should be expecting. Obviously, I'm screwing up in the first place, setting the same parent on both children. Yet,

- IIUC, this is not in line with the docs saying that "there is no intrinsic system within the ORM itself that prevents more than one Child object to be created against the same Parent" because in this case, I do get an error, although not the most explicit/expected one

- at least when the UNIQUE constraint is set (and in fact in both cases), I believe a UNIQUE constraint error would be more suited

I guess in an ideal world, the UNIQUE constraint would be added automatically but if this was easy to achieve it would be done already so the limitation explained above ("no intrinsic system...") is a known shortcoming we must live with. Still I thought I'd open a discussion about this because this had me scratching my head for quite a few hours already.

---------------------------------------

I just found 2 related discussions:

- https://github.com/sqlalchemy/sqlalchemy/discussions/10306
- https://github.com/sqlalchemy/sqlalchemy/discussions/7068

I still think (hope) my points here are relevant as in those discussions, the OP don't mean to create a 1x1 relation IIUC.

**Comment by zzzeek** [comment]
The note refers to the many-to-one side specifically.   your test has differing behaviors only due to the one-to-many side.  however, there is an option that can be used to check the many-to-one side at Python assignment time, which I will add to the documentation note.   We still very much want people setting up unique constraints so we don't want to encourage python-only detection of this condition.

the history system does not detect the two child objects conflicting when they are both pending.  Each child has the same parent as its .parent:

```
(Pdb) child_1.parent
<__main__.Parent object at 0x7fb1739c2ad0>
(Pdb) child_2.parent
<__main__.Parent object at 0x7fb1739c2ad0>
(Pdb) parent_1.child
<__main__.Child object at 0x7fb174bd0b10>
```

however, on the Parent.child side, which is where "there's an old object to remove first before we put the new one in" behavior occurs, this does not happen for two pending objects; history does not track a non-persistent object as "removed", so we only see the one new object added:

```
(Pdb) from sqlalchemy import inspect
(Pdb) inspect(parent_1).attrs.child.history
History(added=[<__main__.Child object at 0x7fb174bd0b10>], unchanged=(), deleted=())
```


when one child/parent is already persistent, the ORM instead sees this:

```
-> session.flush()
(Pdb) from sqlalchemy import inspect
(Pdb) inspect(parent_1).attrs.child.history
History(added=[<__main__.Child object at 0x7fd36a79c650>], unchanged=(), deleted=[<__main__.Child object at 0x7fd36a77fbd0>])
(Pdb) 

```

the entire situation changes if you set single_parent=True on the Child.parent relationship, where a python side error will be raised, so I will add this to the doc now.   we are not ready to recommend this in all cases as it prevents the usual silent replacement operation from succeeding, which might be the behavior that people want, and we also really want folks to place unique constraints on their tables for this case.  the ORM does not like being in the business of doubling what the database constraints already do any more than it has to be.

this new documentation is merged in c03f24e18245f3272e06.

> at least when the UNIQUE constraint is set (and in fact in both cases), I believe a UNIQUE constraint error would be more suited

the setting of child_id=NULL is a core behavior of the ORM that's pretty important.  if you want to turn it off for a special case, set [passive_deletes='all'](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.passive_deletes).
